cpu/sam0_common/periph/i2c.c

```diff
@@ -326,51 +405,111 @@ void _i2c_poweroff(i2c_t dev)
+     *
+     * We ignore the three timeout flags in the status register:
+     * SERCOM_I2CM_STATUS_MEXTTOUT, SERCOM_I2CM_STATUS_LOWTOUT
+     * SERCOM_I2CM_STATUS_SEXTTOUT.
+     *
+     * We further ignore SERCOM_I2CM_STATUS_LENERR (useful in 32 bit
+     * mode and with DMA, not related to address transmission).
+     */
+    uint16_t intflag = dev->INTFLAG.reg;
+    uint16_t status = dev->STATUS.reg;
+
+    /* handle errors */
+    bool unblock_bus = false;
+
+    if (res) {
+        DEBUG_PUTS("i2c.c: STATUS_ERR_TIMEOUT");
+        unblock_bus = true;
     }
-    else {
-        if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
+    else if ((status & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_BUSY) {
+        DEBUG_PUTS("i2c.c: STATUS_ERR_BUS_BUSY");
+        unblock_bus = true;
+        res = -EAGAIN;
     }
-
-    /* Check for address response error unless previous error is detected. */
-    /* Check for error and ignore bus-error; workaround for BUSSTATE
-     * stuck in BUSY */
-    if (dev->INTFLAG.reg & SERCOM_I2CM_INTFLAG_SB) {
-        /* Clear write interrupt flag */
-        dev->INTFLAG.reg = SERCOM_I2CM_INTFLAG_SB;
-        /* Check arbitration. */
-        if (dev->STATUS.reg & SERCOM_I2CM_STATUS_ARBLOST) {
-            DEBUG("STATUS_ERR_PACKET_COLLISION\n");
-            return -EAGAIN;
-        }
+    else if ((status & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN) {
+        DEBUG_PUTS("i2c.c: STATUS_ERR_BUS_STATE_UNKNOWN");
+        unblock_bus = true;
+        res = -EIO;
+    }
+    else if ((intflag & SERCOM_I2CM_INTFLAG_MB) &&
+            (status & SERCOM_I2CM_STATUS_ARBLOST)) {
+        DEBUG_PUTS("i2c.c: STATUS_ERR_ARBLOST");
+        res = -EAGAIN;
+    }
+    else if (status & SERCOM_I2CM_STATUS_BUSERR) {
+        /* a bus error not related to a lost arbitration */
+        DEBUG_PUTS("i2c.c: STATUS_ERR_BUSERR");
+        unblock_bus = true;
+        res = -EIO;
     }
-    /* Check that slave responded with ack. */
-    else if (dev->STATUS.reg & SERCOM_I2CM_STATUS_RXNACK) {
-        /* Slave busy. Issue ack and stop command. */
+    else if ((intflag & SERCOM_I2CM_INTFLAG_MB) &&
+            (status & SERCOM_I2CM_STATUS_RXNACK)) {
+        /* datasheet recommends: send ack + stop condition */
         dev->CTRLB.reg |= SERCOM_I2CM_CTRLB_CMD(3);
-        DEBUG("STATUS_ERR_BAD_ADDRESS\n");
-        return -ENXIO;
+        _syncbusy(dev);
+        DEBUG_PUTS("i2c.c: STATUS_ERR_BUSY_OR_BAD_ADDRESS");
+        res = -ENXIO;
     }
-    return 0;
+
+    if (IS_ACTIVE(ENABLE_DEBUG) && res) {
+        DEBUG("i2c.c: error: res=%d, INTFLAG=0x%04"PRIx16", STATUS=0x%04"PRIx16"\n",
+                res, dev->INTFLAG.reg, dev->STATUS.reg);
+    }
+
+    if (unblock_bus) {
+        _check_and_unblock_bus(dev_id, true);
+    }
+
+    /* Reset flags */
+    dev->INTFLAG.reg = SERCOM_I2CM_INTFLAG_MB & SERCOM_I2CM_INTFLAG_SB &
+        SERCOM_I2CM_INTFLAG_ERROR;
```

Comment:
&ing those will just result in `dev->INTFLAG.reg = 0;`

Reply:
Good catch. Not sure how that slipped in. Resolved in my latest fixup. Any more comments or should I squash?

Reply:
True. That and several others. I added some guards. Does work for me now when test compiling locally with all sam* boards.